larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+
+void HitWidthClusterMergingAlgorithm::CleanupClusterAssociations(const ClusterVector &clusterVector, ClusterAssociationMap &clusterAssociationMap) const
+{
+
+  // Create temporary map so can delete elements whilst still iterating over them
+  ClusterAssociationMap tempMap(clusterAssociationMap);
+
+  for(const Cluster *const pCluster : clusterVector) 
+  {
+      const ClusterAssociationMap::const_iterator primaryMapIter = clusterAssociationMap.find(pCluster);
+
+      if(primaryMapIter == clusterAssociationMap.end())
+          continue;
+
+      const ClusterSet &primaryForwardAssociations(primaryMapIter->second.m_forwardAssociations);
+
+      // loop through the primary associations of each cluster
+      // remove clusters that are present in secondary associations of other primary clusters 
+      for(const Cluster *const pConsideredCluster : primaryForwardAssociations)
+      {
+          for(const Cluster *const pPrimaryCluster : primaryForwardAssociations)
+          {
+              if(pConsideredCluster == pPrimaryCluster)
+                  continue;
+
+              const ClusterAssociationMap::const_iterator secondaryMapIter = clusterAssociationMap.find(pPrimaryCluster);
+
+	          // if primary cluster has no associations (this shouldn't ever be the case)
+              if(secondaryMapIter == clusterAssociationMap.end()) 
+                  continue;
+
+              const ClusterSet &secondaryForwardAssociations(secondaryMapIter->second.m_forwardAssociations);
+
+              // if cluster is present in the forward associations of any cluster at the same level remove
+              if(secondaryForwardAssociations.find(pConsideredCluster) != secondaryForwardAssociations.end()) 
+              {
+                  ClusterSet &tempPrimaryForwardAssociations(tempMap.find(pCluster)->second.m_forwardAssociations);
+                  const ClusterSet::const_iterator forwardAssociationToRemove(tempPrimaryForwardAssociations.find(pConsideredCluster));
+
+                  // if association has already been removed
+                  if(forwardAssociationToRemove == tempPrimaryForwardAssociations.end())
+                      continue;
+
+                  ClusterSet &tempPrimaryBackwardAssociations(tempMap.find(pConsideredCluster)->second.m_backwardAssociations);
+                  const ClusterSet::const_iterator backwardAssociationToRemove(tempPrimaryBackwardAssociations.find(pCluster));
+
+                  // if association has already been removed
+                  if(backwardAssociationToRemove == tempPrimaryBackwardAssociations.end())
+                      continue;
+
+                  tempPrimaryForwardAssociations.erase(forwardAssociationToRemove);
+                  tempPrimaryBackwardAssociations.erase(backwardAssociationToRemove);
+              }
+          }
+      }
+  }
+  clusterAssociationMap = tempMap;
```

Comment:
This looks good, from reading over it (with comments about use of ClusterSet). Obviously we'll soak test too.